### PR TITLE
lint against `MaybeUninit::uninit().assume_init()`

### LINKED ISSUE #4272: Lint against MaybeUninit::uninit().assume_init()
I spotted this in PyO3: (https://github.com/PyO3/pyo3/issues/536)

```rust
if next_idx == 0 {
    self.inner
        .push_back(unsafe { mem::MaybeUninit::uninit().assume_init() });
}
```

GitHub search brings up a total of 62 matches even for a fairly restrictive search query: https://github.com/search?q=%22uninit%28%29%3A%3Aassume_init%28%29%22&type=Code

**Problem:** It seems people are using this as an easy way out of the deprecation warnings for `std::mem::uninitialized`, failing to understand the reason *why* it was deprecated (it is, to my current understanding, universally UB, *always!*).  The correct solution is to use `MaybeUninit<T>` where you were formerly using T until the value is known to be initialized.  The `MaybeUninit<T>` docs even give examples for what are by far the two most common use cases (out pointers and partially-initialized arrays).

**Proposal:** There should either be a warn-by-default or an error-by-default lint against immediate calls to `assume_init()` on `MaybeUninit::uninit()` with no other intervening operations.

**Possible extension:** We can try going a step further and adjust our suggestions based on the type:

* In the case where `T` is uninhabited, we can instead recommend the use of `std::hint::unreachable_unchecked`.
* If `T = [A; N]`, we can specifically recommend that the user constructs a `[MaybeUninit<A>; N]`. (as this is often a key piece of the correct solution that the user might not think of on their own, and without it they may simply invent some other way to silence the warning improperly)

### PR DESCRIPTION
**Pull request by llogiq**
changelog: add `uninit_assumed_init` lint

This fixes #4272

### CODE REVIEW

File: clippy_lints/src/methods/mod.rs [hunk 1]
```diff
@@ -951,6 +951,38 @@ declare_clippy_lint! {
...
+    /// of `MaybeUninit`. There may be other types that allow uninitialized
+    /// data, but those are not yet rigorously defined.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// // Beware the UB
+    /// use std::mem::MaybeUninit;
+    ///
+    /// let _: usize = unsafe { MaybeUninit::uninit().assume_init() };
```
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs]: Is there a correct way to do the same? Or ist this just a "Don't do this at all" lint?

File: clippy_lints/src/methods/mod.rs [hunk 2]
```diff
@@ -2662,6 +2696,37 @@ fn lint_into_iter(cx: &LateContext<'_, '_>, expr: &hir::Expr, self_ref_ty: Ty<'_
...
+        if args.is_empty();
+        if let hir::ExprKind::Path(ref path) = callee.node;
+        if match_qpath(path, &paths::MEM_MAYBEUNINIT_UNINIT);
+        if !check_maybe_uninit_ty(cx, cx.tables.expr_ty_adjusted(outer));
+        then {
+            span_lint(
+                cx,
+                UNINIT_ASSUMED_INIT,
+                outer.span,
+                "This call for this type may be undefined behavior"
```
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs]: NIT:
```suggestion
                "this call for this type may be undefined behavior"
```

File: clippy_lints/src/methods/mod.rs (hunk 1 above)
Comment (llogiq) [reply clippy_lints/src/methods/mod.rs]: The correct way is hard to know in the general case. Perhaps for `Default` types we could suggest using `Default::default` instead. But that might be misleading if the user made a type error and wanted to initialize a uninit-compatible type instead.

File: clippy_lints/src/methods/mod.rs (hunk 2 above)
Comment (llogiq) [reply clippy_lints/src/methods/mod.rs]: Thanks, added.

**Comment by llogiq** [comment]
For now, this is a "Don't do this" lint. Working with uninitialized values is so complex that I cannot come up with an easy suggestion that would be sound in all cases.

@flip1995 r?

**Comment by flip1995** [comment]
Yeah I did some reading on this topic yesterday. A suggestion that applies in general is probably not possible.

Since Clippy broke on the latest rustc master and there wasn't a rustup yet, we have to wait on merging this. cc rust-lang/rust#63561

File: clippy_lints/src/methods/mod.rs [hunk 3]
```diff
@@ -2662,6 +2696,37 @@ fn lint_into_iter(cx: &LateContext<'_, '_>, expr: &hir::Expr, self_ref_ty: Ty<'_
...
+    }
+}
+
+fn check_maybe_uninit_ty(cx: &LateContext<'_, '_>, ty: Ty<'_>) -> bool {
+    match ty.sty {
+        ty::Array(ref component, _) => check_maybe_uninit_ty(cx, component),
+        ty::Tuple(ref types) => types.types().all(|ty| check_maybe_uninit_ty(cx, ty)),
+        ty::Adt(ref adt, _) => {
+            // needs to be a MaybeUninit
+            match_def_path(cx, adt.did, &paths::MEM_MAYBEUNINIT)
```
Comment (eddyb) [inline clippy_lints/src/methods/mod.rs]: Isn't `MaybeUninit` a lang item?

File: clippy_lints/src/methods/mod.rs [hunk 4]
```diff
@@ -2662,6 +2696,37 @@ fn lint_into_iter(cx: &LateContext<'_, '_>, expr: &hir::Expr, self_ref_ty: Ty<'_
...
+                cx,
+                UNINIT_ASSUMED_INIT,
+                outer.span,
+                "this call for this type may be undefined behavior"
+            );
+        }
+    }
+}
+
+fn check_maybe_uninit_ty(cx: &LateContext<'_, '_>, ty: Ty<'_>) -> bool {
```
Comment (eddyb) [inline clippy_lints/src/methods/mod.rs]: This should probably be called `check_ty_valid_even_if_uninit` or something like that.

Comment (llogiq) [reply clippy_lints/src/methods/mod.rs]: renamed it to `is_maybe_uninit_ty_valid`.

File: clippy_lints/src/methods/mod.rs (hunk 3 above)
Comment (llogiq) [reply clippy_lints/src/methods/mod.rs]: No. `ManuallyDrop` is, but `MaybeUninit` isn't.

**Comment by flip1995** [comment]
Thanks!

@bors r+

**Pull request merged**